Draft-Accepted/RFCXXXX-Using-IDisposables.md

```diff
@@ -0,0 +1,399 @@
+---
+RFC: RFCNNNN-Using-IDisposables.md
+Author: Jordan Borean
+Status: Draft
+SupercededBy:
+Version: 1.0
+Area: engine
+Comments Due: July 1, 2024
+Plan to implement: Yes
+---
+
+# Using Statement Blocks for IDisposables
+
+## Motivation
+
+    As a powershell developer/user,
+    I can enclose my disposable instances in a block,
+    so that they are disposed automatically when they are no longer needed.
+
+## User Experience
+
+PowerShell can interact with many instance types that might hold onto system resources which are tied to the lifetime of that instance.
+Examples of such types are the [FileStream](https://learn.microsoft.com/en-us/dotnet/api/system.io.filestream?view=net-8.0), [RegistryKey](https://learn.microsoft.com/en-us/dotnet/api/microsoft.win32.registrykey?view=net-8.0), [Process](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process?view=net-8.0) which contain operating system (OS) handles/identifiers.
+On Windows, some of these handles could place a lock which can impact future operations, like trying to open a new handle on a file locked for reading, unload a registry hive with a key still open, etc.
+
+A safe way to deal with these objects is to wrap the code that creates the instance in a `try/finally` block.
+For example the below PowerShell script interacts with various streams that should be disposed when no longer needed to free up OS resources or internal memory buffers:
+
+```powershell
+$reader = [System.IO.File]::OpenText("numbers.txt")
+try {
+    [Console]::WriteLine($reader.ReadLine())
+}
+finally {
+    $reader.Dispose()
+}
+```
+
+The drawbacks here are
+
++ The user needs to use a `try/finally` block to ensure it is cleaned up
++ The `finally` block can be interrupted with an engine stop request
++ The user need to explicitly write code not only to pre-define the vars, set the vars, then dispose of them with a null check
+
+C# developers can use the [using statement](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/using) that allows them to scope an `IDisposable` instance in an automatic `try/finally` block like the above:
+
+```csharp
+using (StreamReader reader = File.OpenText("numbers.txt"))
+{
+    Console.WriteLine(reader.ReadLine());
+} // reader is disposed here
+```
+
+This RFC proposes adding an equivalent `using` statement in PowerShell that can do the same thing as C#.
+For example the same C# `StreamReader` example in PowerShell can now be:
+
```

Comment:
> there may be some collisions with implementations of a `using` function (I have one in my profile), so I think this would qualify as a breaking change.

`using` is thankfully already a keyword, so you wouldn't be able to use a command named `using` without `& 'using'` otherwise you'd get a parse error. `use` is a common alias/function for this, but should be unaffected

Reply:
Yea the fact that `using` is already a keyword is another reason why I chose to use it over something like `dispose`/`clean`. People can't do `using () {}` as a command without the call operator so it's a nice way to avoid a breaking change.